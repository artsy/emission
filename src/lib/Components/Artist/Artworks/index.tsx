import * as React from "react"
import { StyleSheet, View, ViewProperties } from "react-native"
import * as Relay from "react-relay"

import ArtworksGrid from "../../artwork_grids/relay_connections/artist_artworks_grid"
import Separator from "../../separator"
import SerifText from "../../text/serif"

import colors from "../../../../data/colors"

interface Props extends ViewProperties {
  artist: {
    counts: {
      for_sale_artworks: number
      artworks: number
    }
    not_for_sale_artworks: any[]
    for_sale_artworks: any[]
  }
  relay: any
}

interface State {
  completedForSaleWorks: boolean
}

class Artworks extends React.Component<Props, State> {
  constructor(props) {
    super(props)
    this.state = {
      completedForSaleWorks: false,
    }
  }

  render() {
    const forSaleCount = this.props.artist.counts.for_sale_artworks
    const otherCount = this.props.artist.counts.artworks - forSaleCount
    if (forSaleCount === 0) {
      return this.renderSection({
        title: "Works",
        count: otherCount,
        filter: "IS_NOT_FOR_SALE",
        onComplete: null,
      })
    } else {
      const otherWorks: any[] = []
      const showOtherWorks = otherCount > 0 && (forSaleCount < 10 || this.state.completedForSaleWorks)
      if (showOtherWorks) {
        otherWorks.push(<Separator style={styles.sectionSeparator} key="separator" />)
        otherWorks.push(
          this.renderSection({
            title: "Other Works",
            count: otherCount,
            filter: "IS_NOT_FOR_SALE",
            onComplete: null,
          })
        )
      }
      return (
        <View style={styles.section}>
          {this.renderSection({
            title: "Works for Sale",
            count: forSaleCount,
            filter: "IS_FOR_SALE",
            onComplete: () => {
              this.setState({ completedForSaleWorks: true })
            },
          })}
          {otherWorks}
        </View>
      )
    }
  }

  renderSection({ title, count, filter, onComplete }) {
    const countStyles = [styles.text, styles.count]
    return (
      <View key={title}>
        <SerifText style={styles.heading}>
          <SerifText style={styles.text}>{title}</SerifText> <SerifText style={countStyles}>({count})</SerifText>
        </SerifText>
        <ArtworksGrid artist={this.props.artist} filter={filter} onComplete={onComplete} queryKey="artist" />
      </View>
    )
  }
}

const styles = StyleSheet.create({
  heading: {
    marginBottom: 20,
  },
  text: {
    fontSize: 20,
  },
  count: {
    color: colors["gray-semibold"],
  },
  section: {
    marginBottom: 40,
  },
  sectionSeparator: {
    marginTop: 40, // FIXME: This is because the above `section.marginBottom` didnâ€™t work before the separator.
    marginBottom: 20,
  },
})

export default Relay.createContainer(Artworks, {
  fragments: {
    artist: () => Relay.QL`
      fragment on Artist {
        counts {
          artworks
          for_sale_artworks
        }
        ${ArtworksGrid.getFragment("artist", { filter: "IS_FOR_SALE" })}
        ${ArtworksGrid.getFragment("artist", { filter: "IS_NOT_FOR_SALE" })}
      }
    `,
  },
})

interface RelayProps {
  artist: {
    counts: {
      artworks: boolean | number | string | null
      for_sale_artworks: boolean | number | string | null
    } | null
    for_sale_artworks: Array<boolean | number | string | null> | null
    not_for_sale_artworks: Array<boolean | number | string | null> | null
  }
}
