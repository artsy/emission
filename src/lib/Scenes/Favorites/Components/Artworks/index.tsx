import React, { Component } from "react"
import { createPaginationContainer, graphql, RelayPaginationProp } from "react-relay"
import styled from "styled-components/native"

import GenericGrid from "lib/Components/ArtworkGrids/GenericGrid"
import ZeroState from "lib/Components/States/ZeroState"
import { PAGE_SIZE } from "lib/data/constants"
import { isCloseToBottom } from "lib/utils/isCloseToBottom"

const Container = styled.ScrollView`
  padding: 20px;
  flex: 1;
`

interface Props extends RelayProps {
  relay?: RelayPaginationProp
  onDataFetching?: (loading: boolean) => void
}

export class SavedWorks extends Component<Props> {
  loadMore = () => {
    if (!this.props.relay.hasMore() || this.props.relay.isLoading()) {
      return
    }

    const updateState = (loading: boolean) => {
      this.setState({ fetchingMoreData: loading })
      if (this.props.onDataFetching) {
        this.props.onDataFetching(loading)
      }
    }

    updateState(true)
    this.props.relay.loadMore(PAGE_SIZE, error => {
      if (error) {
        // FIXME: Handle error
        console.error("Artworks/index.tsx", error.message)
      }
      updateState(false)
    })
  }

  render() {
    const artworks = this.props.me.saved_artworks.artworks_connection.edges.map(edge => edge.node)

    const EmptyState = (
      <ZeroState
        title="You havenâ€™t followed any artists yet"
        subtitle="Follow artists to get notified about new works that have been added to Artsy."
      />
    )

    const Content = (
      <Container onScroll={isCloseToBottom(this.loadMore)} scrollEventThrottle={400}>
        <GenericGrid artworks={artworks} />
      </Container>
    )

    return artworks.length ? Content : EmptyState
  }
}

export default createPaginationContainer(
  SavedWorks,
  {
    me: graphql`
      fragment Artworks_me on Me
        @argumentDefinitions(count: { type: "Int", defaultValue: 10 }, cursor: { type: "String", defaultValue: "" }) {
        saved_artworks {
          artworks_connection(private: true, first: $count, after: $cursor)
            @connection(key: "GenericGrid_artworks_connection") {
            pageInfo {
              endCursor
              hasNextPage
            }
            edges {
              node {
                ...GenericGrid_artworks
              }
            }
          }
        }
      }
    `,
  },
  {
    direction: "forward",
    getConnectionFromProps(props) {
      return props.me && props.me.saved_artworks.artworks_connection
    },
    getFragmentVariables(prevVars, totalCount) {
      return {
        ...prevVars,
        count: totalCount,
      }
    },
    getVariables(_props, { count, cursor }, fragmentVariables) {
      return {
        ...fragmentVariables,
        count,
        cursor,
      }
    },
    query: graphql`
      query ArtworksQuery($count: Int!, $cursor: String) {
        me {
          ...Artworks_me @arguments(count: $count, cursor: $cursor)
        }
      }
    `,
  }
)

interface RelayProps {
  me: {
    saved_artworks: {
      artworks_connection: {
        edges: Array<{
          node: any | null
        }> | null
      } | null
    } | null
  } | null
}
