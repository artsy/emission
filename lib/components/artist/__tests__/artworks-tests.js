import 'react-native'

import React from 'react'
import renderer from 'react-test-renderer'

import Artworks from '../artworks'
jest.mock('../../artwork_grids/infinite_scroll_grid.js', () => 'ArtworksGrid')

it('renders properly', () => {
  const artist = {
    counts: {
      artworks: 5,
      for_sale_artworks: 2,
    },
  }
  const artworks = renderer.create(<Artworks artist={artist} />).toJSON()
  expect(artworks).toMatchSnapshot()
})
