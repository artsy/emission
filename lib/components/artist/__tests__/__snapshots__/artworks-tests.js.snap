exports[`test renders properly 1`] = `
<View
  style={
    Object {
      "marginBottom": 40,
    }
  }>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      numberOfLines={1}
      style={
        Array [
          Object {
            "fontSize": 17,
          },
          Object {
            "marginBottom": 20,
          },
          Object {
            "fontFamily": "AGaramondPro-Regular",
          },
        ]
      }>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontSize": 20,
            },
            Object {
              "fontFamily": "AGaramondPro-Regular",
            },
          ]
        }>
        Works for Sale
      </Text>
       
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Array [
              Object {
                "fontSize": 20,
              },
              Object {
                "color": "#666666",
              },
            ],
            Object {
              "fontFamily": "AGaramondPro-Regular",
            },
          ]
        }>
        (
        2
        )
      </Text>
    </Text>
    <ArtworksGrid
      artist={
        Object {
          "counts": Object {
            "artworks": 5,
            "for_sale_artworks": 2,
          },
        }
      }
      filter="IS_FOR_SALE"
      onComplete={[Function]} />
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#e5e5e5",
          "height": 1,
          "marginLeft": -20,
          "marginRight": -20,
        },
        Object {
          "marginBottom": 20,
          "marginTop": 40,
        },
      ]
    } />
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      numberOfLines={1}
      style={
        Array [
          Object {
            "fontSize": 17,
          },
          Object {
            "marginBottom": 20,
          },
          Object {
            "fontFamily": "AGaramondPro-Regular",
          },
        ]
      }>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Object {
              "fontSize": 20,
            },
            Object {
              "fontFamily": "AGaramondPro-Regular",
            },
          ]
        }>
        Other Works
      </Text>
       
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        numberOfLines={1}
        style={
          Array [
            Object {
              "fontSize": 17,
            },
            Array [
              Object {
                "fontSize": 20,
              },
              Object {
                "color": "#666666",
              },
            ],
            Object {
              "fontFamily": "AGaramondPro-Regular",
            },
          ]
        }>
        (
        3
        )
      </Text>
    </Text>
    <ArtworksGrid
      artist={
        Object {
          "counts": Object {
            "artworks": 5,
            "for_sale_artworks": 2,
          },
        }
      }
      filter="IS_NOT_FOR_SALE"
      onComplete={null} />
  </View>
</View>
`;
